include/bluetooth/services/cts_client.h

```diff
@@ -0,0 +1,152 @@
+struct bt_cts_exact_time_256 {
+	uint16_t year;
+	uint8_t month;
+	uint8_t day;
+	uint8_t hours;
+	uint8_t minutes;
+	uint8_t seconds;
+	uint8_t day_of_week;
+	uint8_t fractions256;
+};
+
+/**@brief "Adjust Reason" field of the Current Time characteristic. */
+struct bt_cts_adjust_reason {
+	uint8_t manual_time_update : 1;
+	uint8_t external_reference_time_update : 1;
+	uint8_t change_of_time_zone : 1;
+	uint8_t change_of_daylight_savings_time : 1;
+};
+
+/**@brief Data structure for the Current Time characteristic. */
+struct bt_cts_current_time_char {
+	struct bt_cts_exact_time_256 exact_time_256;
+	struct bt_cts_adjust_reason adjust_reason;
+};
+
+/**@brief Current Time Service client event type. */
+enum bt_cts_client_evt_type {
+	/** A new Current Time reading has been received. */
+	BT_CTS_CLIENT_EVT_CURRENT_TIME,
+
+	/** The Current Time value received from the peer is invalid. */
+	BT_CTS_CLIENT_EVT_INVALID_TIME
+};
+
+/**@brief Current Time Service client event. */
+struct bt_cts_client_evt {
+	/** Type of event. */
+	enum bt_cts_client_evt_type evt_type;
+
+	/** Current Time characteristic data. This is filled when the evt_type is @ref BT_CTS_CLIENT_EVT_CURRENT_TIME. */
+	struct bt_cts_current_time_char current_time;
+};
+
+struct bt_cts_client;
+
+/**@brief Current Time Service client event handler type. */
+typedef void (*bt_cts_client_evt_handler)(struct bt_cts_client *cts_c,
+					  const struct bt_cts_client_evt *evt);
+
+/**@brief Current Time Service client structure. This structure contains status information for the client. */
+struct bt_cts_client {
+	/** Connection object. */
+	struct bt_conn *conn;
+
+	/** Handle of the Current Time Characteristic. */
+	uint16_t handle_ct;
+
+	/** Handle of the CCCD of the Current Time Characteristic. */
+	uint16_t handle_ct_ccc;
+
+	/** GATT read parameters. */
+	struct bt_gatt_read_params read_params;
+
+	/** Semaphore for GATT read parameters. */
+	struct k_sem read_params_sem;
+
+	/** Event handler to be called for handling events from the Current Time Service client. */
+	bt_cts_client_evt_handler evt_handler;
+};
+
+/**@brief Function for initializing the Current Time Service client.
+ *
+ * @param[out] cts_c       Current Time Service client structure. This structure must
+ *                         be supplied by the application. It is initialized by this
+ *                         function and can later be used to identify this particular client
+ *                         instance.
+ * @param[in]  evt_handler Event handler for Current Time Service client.
+ *
+ * @retval 0 If the client was initialized successfully.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int bt_cts_client_init(struct bt_cts_client *cts_c,
+		       bt_cts_client_evt_handler evt_handler);
+
+/**@brief Function for clearing the CTS client connection.
+ *
+ * @param[in] cts_c  Current Time Service client structure.
+ */
+void bt_cts_client_on_disconnected(struct bt_cts_client *cts_c);
```

Comment:
It would be better for the CTS client module to register it's own callback instead of requiring the application to call this function (see BMS)

Possibly this could be removed entirely by removing the need for the semaphore.

Reply:
The bt_conn_cb_register() callbacks have bt_conn structure as parameter. It is not convenient for the CTS client module to get the bt_cts_client structure.

The BAS client tackles this issue by bas_reinit(). It clears the flags on assigning GATT handles. I will follow suit.